smartsim/_core/mli/workermanager.py

```diff
@@ -0,0 +1,976 @@
+        cooldown: int = 0,
+    ) -> None:
+        """Initialize the WorkerManager
+        :param feature_store: The persistence mechanism
+        :param worker: A worker to manage
+        :param as_service: Specifies run-once or run-until-complete behavior of service
+        :param cooldown: Number of seconds to wait before shutting down afer
+        shutdown criteria are met"""
+        super().__init__(as_service, cooldown)
+
+        self._workers: t.Dict[
+            str, "t.Tuple[MachineLearningWorkerBase, mp.Queue[bytes]]"
+        ] = {}
+        """a collection of workers the manager is controlling"""
+        self._upstream_queue: t.Optional[mp.Queue[bytes]] = None
+        """the queue the manager monitors for new tasks"""
+        self._feature_store: FeatureStore = feature_store
+        """a feature store to retrieve models from"""
+        self._worker = worker
+        """The ML Worker implementation"""
+
+    @property
+    def upstream_queue(self) -> "t.Optional[mp.Queue[bytes]]":
+        """Return the queue used by the worker manager to receive new work"""
+        return self._upstream_queue
+
+    @upstream_queue.setter
+    def upstream_queue(self, value: "mp.Queue[bytes]") -> None:
+        """Set/update the queue used by the worker manager to receive new work"""
+        self._upstream_queue = value
+
+    def _on_iteration(self, timestamp: int) -> None:
+        """Executes calls to the machine learning worker implementation to complete
+        the inference pipeline"""
+        logger.debug(f"{timestamp} executing worker manager pipeline")
+
+        if self.upstream_queue is None:
+            logger.warning("No queue to check for tasks")
+            return
+
+        msg: bytes = self.upstream_queue.get()
+        request = self._worker.deserialize(msg)
+        fetch_model_result = self._worker.fetch_model(request, self._feature_store)
+        model_result = self._worker.load_model(request, fetch_model_result)
+        fetch_input_result = self._worker.fetch_inputs(
+            request,
+            self._feature_store,
+        )  # we don't know if they'lll fetch in some weird way
+        # they potentially need access to custom attributes
+        # we don't know what the response really is... i have it as bytes
+        # but we just want to advertise that the contract states "the output
+        # will be the input to transform_input... "
+
+        transform_result = self._worker.transform_input(request, fetch_input_result)
+
+        # batch: t.Collection[_Datum] = transform_result.transformed_input
+        # if self._batch_size:
+        #     batch = self._worker.batch_requests(transform_result, self._batch_size)
+
+        # todo: what do we return here? tensors? Datum? bytes?
+        results = self._worker.execute(request, model_result, transform_result)
+
+        # todo: inference reply _must_ be replaced with the mli schemas reply
+        reply = InferenceReply()
```

Comment:
@al-rigazzi For this type is it problematic to label it with ``Inference`` instead of something more generic?  Trauma from the over utilization of "launcher" made me question this.

Reply:
I think it is usually something along the lines of `Prediction`, but honestly `Inference` sounds better and more generic

Reply:
Per convo, leaving as inference

Reply:
per offline, keeping `Inference`